sound/soc/sof/amd/acp-stream.c

```diff
@@ -0,0 +1,210 @@
+#include <linux/err.h>
+#include <sound/pcm_params.h>
+#include <linux/module.h>
+#include <sound/sof.h>
+#include <sound/sof/xtensa.h>
+#include "../sof-priv.h"
+#include "../sof-audio.h"
+
+#include "../ops.h"
+#include "acp.h"
+
+#define PTE_GRP1_OFFSET		0x00000000
+#define PTE_GRP2_OFFSET		0x00800000
+#define PTE_GRP3_OFFSET		0x01000000
+#define PTE_GRP4_OFFSET		0x01800000
+#define PTE_GRP5_OFFSET		0x02000000
+#define PTE_GRP6_OFFSET		0x02800000
+#define PTE_GRP7_OFFSET		0x03000000
+#define PTE_GRP8_OFFSET		0x03800000
+
+int config_pte_for_stream(struct snd_sof_dev *sdev,
+			  struct acp_dsp_stream *stream, int num_pages)
+{
+	struct snd_pcm_substream *substream = stream->substream;
+	int stream_tag = stream->stream_tag;
+	struct snd_dma_buffer *dmab = substream->runtime->dma_buffer_p;
+	unsigned int pte_reg = 0, pte_size = 0;
+	unsigned int phy_addr_offset = 0;
+	unsigned int index = 0;
+	u32 low, high, offset = 0, reg_val = 0;
+	dma_addr_t addr = 0;
+	int page_idx;
+
+	switch (stream_tag) {
+	case 1:
+		pte_reg = ACPAXI2AXI_ATU_BASE_ADDR_GRP_1;
+		pte_size =  ACPAXI2AXI_ATU_PAGE_SIZE_GRP_1;
+		offset = offsetof(struct acp_scratch_register_config,
+				  acp_atu_grp1_pte);
+		stream->reg_offset = PTE_GRP1_OFFSET;
+		break;
+	case 2:
+		pte_reg = ACPAXI2AXI_ATU_BASE_ADDR_GRP_2;
+		pte_size =  ACPAXI2AXI_ATU_PAGE_SIZE_GRP_2;
+		offset = offsetof(struct acp_scratch_register_config,
+				  acp_atu_grp2_pte);
+		stream->reg_offset = PTE_GRP2_OFFSET;
+		break;
+	case 3:
+		pte_reg = ACPAXI2AXI_ATU_BASE_ADDR_GRP_3;
+		pte_size =  ACPAXI2AXI_ATU_PAGE_SIZE_GRP_3;
+		offset = offsetof(struct acp_scratch_register_config,
+				  acp_atu_grp3_pte);
+		stream->reg_offset = PTE_GRP3_OFFSET;
+		break;
+	case 4:
+		pte_reg = ACPAXI2AXI_ATU_BASE_ADDR_GRP_4;
+		pte_size =  ACPAXI2AXI_ATU_PAGE_SIZE_GRP_4;
+		offset = offsetof(struct acp_scratch_register_config,
+				  acp_atu_grp4_pte);
+		stream->reg_offset = PTE_GRP4_OFFSET;
+		break;
+	case 5:
+		pte_reg = ACPAXI2AXI_ATU_BASE_ADDR_GRP_5;
+		pte_size =  ACPAXI2AXI_ATU_PAGE_SIZE_GRP_5;
+		offset = offsetof(struct acp_scratch_register_config,
+				  acp_atu_grp5_pte);
+		stream->reg_offset = PTE_GRP5_OFFSET;
+		break;
+	case 6:
+		pte_reg = ACPAXI2AXI_ATU_BASE_ADDR_GRP_6;
+		pte_size =  ACPAXI2AXI_ATU_PAGE_SIZE_GRP_6;
+		offset = offsetof(struct acp_scratch_register_config,
+				  acp_atu_grp6_pte);
+		stream->reg_offset = PTE_GRP6_OFFSET;
+		break;
+	case 7:
+		pte_reg = ACPAXI2AXI_ATU_BASE_ADDR_GRP_7;
+		pte_size =  ACPAXI2AXI_ATU_PAGE_SIZE_GRP_7;
+		offset = offsetof(struct acp_scratch_register_config,
+				  acp_atu_grp7_pte);
+		stream->reg_offset = PTE_GRP7_OFFSET;
+		break;
+	case 8:
+		pte_reg = ACPAXI2AXI_ATU_BASE_ADDR_GRP_8;
+		pte_size =  ACPAXI2AXI_ATU_PAGE_SIZE_GRP_8;
+		offset = offsetof(struct acp_scratch_register_config,
+				  acp_atu_grp8_pte);
+		stream->reg_offset = PTE_GRP8_OFFSET;
+		break;
+	default:
+		pr_err("Invalid stream tag %d\n", stream_tag);
+		break;
```

Comment:
if this is an error, why proceed and return 0? and why pr_err, dev_err(sdev->dev...) maybe?